src/System.IO.Pipes/src/System/IO/Pipes/PipeCompletionSource.cs

```diff
@@ -0,0 +1,181 @@
+                Debug.Assert(completionSource.Overlapped == pOverlapped);
+
+                completionSource.AsyncCallback(errorCode, numBytes);
+            }, this, pinData);
+        }
+
+        internal NativeOverlapped* Overlapped
+        {
+            [SecurityCritical]get { return _overlapped; }
+        }
+
+        protected int ErrorCode { get; private set; }
+
+        internal void RegisterForCancellation()
+        {
+#if DEBUG
+            Debug.Assert(!_cancellationHasBeenRegistered, "Cannot register for cancellation twice");
+            _cancellationHasBeenRegistered = true;
+#endif
+
+            // Quick check to make sure that the cancellation token supports cancellation, and that the IO hasn't completed
+            if (_cancellationToken.CanBeCanceled && Overlapped != null)
+            {
+                // Register the cancellation only if the IO hasn't completed
+                int state = Interlocked.CompareExchange(ref _state, RegisteringCancellation, NoResult);
+                if (state == NoResult)
+                {
+                    // Register the cancellation
+                    _cancellationRegistration = _cancellationToken.Register(thisRef => ((PipeCompletionSource<TResult>)thisRef).Cancel(), this);
+
+                    // Grab the state for case if IO completed while we were setting the registration.
+                    state = Interlocked.Exchange(ref _state, NoResult);
+                }
+                else if (state != CompletedCallback)
+                {
+                    // IO already completed and we have grabbed result state.
+                    // Set NoResult to prevent invocation of CompleteCallback(result state) from AsyncCallback(...)
+                    state = Interlocked.Exchange(ref _state, NoResult);
+                }
+
+                // If we have the result state of completed IO call CompleteCallback(result).
+                // Otherwise IO not completed.
+                if (state == ResultSuccess || state == ResultError)
+                {
+                    CompleteCallback(state);
+                }
+            }
+        }
+
+        internal void ReleaseResources()
+        {
+            _cancellationRegistration.Dispose();
+
+            // NOTE: The cancellation must *NOT* be running at this point, or it may observe freed memory
+            // (this is why we disposed the registration above)
+            if (Overlapped != null)
+            {
+                _threadPoolBinding.FreeNativeOverlapped(Overlapped);
+                _overlapped = null;
+            }
+        }
+
+        internal abstract void SetCompletedSynchronously();
+
+        protected virtual void AsyncCallback(uint errorCode, uint numBytes)
+        {
+            ErrorCode = (int)errorCode;
+
+            int resultState = errorCode == 0 ? ResultSuccess : ResultError;
```

Comment:
These two lines could instead be:
```C#
if (errorCode == 0)
{
    resultState = ResultSuccess;
}
else
{
    ErrorCode = (int)errorCode;
    resultState = ResultError;
}
```
Then we don't have to store the error code if we'd just be overwriting 0 with 0.